File: PP-Lab6/Lab6/F3.cpp
#include "F3.h"

F3::F3(int N) {
	this->N = N;
}

DWORD F3::run() {
	cout << "T3 started." << endl;
	Data* data = new Data(N);

	int t;
	vector<int> V, O, P;
	vector<vector<int>> MO, MP, MR;

	// Generate Input Values
	cout << "T3 gets a permit." << endl;

	t = 3;
	V = data->FillVectorWithNumber(3);
	O = data->FillVectorWithNumber(3);
	P = data->FillVectorWithNumber(3);
	MO = data->FillMatrixWithNumber(3);
	MP = data->FillMatrixWithNumber(3);
	MR = data->FillMatrixWithNumber(3);

	// Calculate The Result
	vector<int> result = data->Func3(t, V, O, P, MO, MP, MR);
	Sleep(100);

	// Output
	if (N <= 10) {
		cout << "T3 result:\n";
		data->VectorOutput(result, 'S');
	}

	cout << "T3 releases the permit." << endl;
	cout << "T3 finished.\n" << endl;

	delete data;

	return 0;
}

File: PP-Lab3/Lab3/F3.cs
using System;
using System.Threading;
using static Lab3.Lab3;

namespace Lab3
{
    public class F3
    {
        // S = (MO*MP)*V+t*MR*(O+P)
        public void Run()
        {
            Console.WriteLine("T3 started.");
            Data data = new Data(N);

            int t;
            int[] V, O, P;
            int[ , ] MO, MP, MR;

            // Generate Input Values
            Thread.Sleep(50);
            Console.WriteLine("T3 is waiting for a permit.");

            Lab3.sem.WaitOne();
            Thread.Sleep(100);
            Console.WriteLine("\nT3 gets a permit.\n");

            t = data.NumInput('t');
            V = data.VectorInput('V'); O = data.VectorInput('O'); P = data.VectorInput('P');
            MO = data.MatrixInput("MO"); MP = data.MatrixInput("MP"); MR = data.MatrixInput("MR");
                
            Console.WriteLine("\nT3 releases the permit.");
            Lab3.sem.Release();
            Console.WriteLine("\nT3 is waiting for a permit.");

            // Calculate The Result
            int[] result = data.Func3(t, V, O, P, MO, MP, MR);
            Thread.Sleep(100);


            // Output
            Lab3.sem.WaitOne();

            Console.WriteLine("T3 gets a permit.\n");
            Console.Write("T3 result:\n");
            data.VectorOutput(result, 'S');

            Console.WriteLine("T3 releases the permit.");
            Lab3.sem.Release();

            Console.WriteLine("T3 finished.\n");
        }
    }

}
